### PR TITLE
Add string conversion error modes

### PR DESCRIPTION
**Pull request by dcodeIO**
Adds a new `ErrorMode` option to UTF-8 encoding, allowing to either keep unpaired surrogate (default, as of WTF-8), replace them (lossy to UTF-8), or error. Defaults remain unchanged.

- [x] I've read the contributing guidelines

### CODE REVIEW

**Review by MaxGraey — APPROVED** [review]

File: lib/loader/index.js [hunk 1]
```diff
@@ -37,17 +37,26 @@ const ARRAY_SIZE = 16;
 const BIGINT = typeof BigUint64Array !== "undefined";
 const THIS = Symbol();
 
-const STRING_DECODE_THRESHOLD = 32;
-const decoder = new TextDecoder("utf-16le");
+const STRING_SMALLSIZE = 192; // break-even point in V8
+const STRING_CHUNKSIZE = 1024; // mitigate stack overflow
+const utf16 = new TextDecoder("utf-16le"); // != wtf16
```
Comment (MaxGraey) [inline lib/loader/index.js:42]:
```suggestion
const utf16 = new TextDecoder("utf-16le", { fatal: true }); // != wtf16
```

File: lib/loader/index.js [hunk 2]
```diff
@@ -37,17 +37,26 @@ const ARRAY_SIZE = 16;
...
 function getStringImpl(buffer, ptr) {
-  const len = new Uint32Array(buffer)[ptr + SIZE_OFFSET >>> 2] >>> 1;
-  const arr = new Uint16Array(buffer, ptr, len);
-  if (len <= STRING_DECODE_THRESHOLD) {
-    return String.fromCharCode.apply(String, arr);
+  let len = new Uint32Array(buffer)[ptr + SIZE_OFFSET >>> 2] >>> 1;
+  const wtf16 = new Uint16Array(buffer, ptr, len);
+  if (len <= STRING_SMALLSIZE) return String.fromCharCode.apply(String, wtf16);
+  try {
+    return utf16.decode(wtf16, { fatal: true });
```
Comment (MaxGraey) [inline lib/loader/index.js:50]:
```suggestion
    return utf16.decode(wtf16);
```

File: lib/loader/umd/index.js [hunk 3]
```diff
@@ -44,19 +44,34 @@ var loader = (function(exports) {
...
-    if (len <= STRING_DECODE_THRESHOLD) {
-      return String.fromCharCode.apply(String, arr);
-    }
+    try {
+      return utf16.decode(wtf16, {
+        fatal: true
+      });
+    } catch {
+      let str = "";
+      let off = 0;
```
Comment (MaxGraey) [inline lib/loader/umd/index.js:66]:
```suggestion
      let str = "", off = 0;
```

File: lib/loader/index.js [hunk 4]
```diff
@@ -37,17 +37,24 @@ const ARRAY_SIZE = 16;
...
-/** Gets a string from an U32 and an U16 view on a memory. */
+/** Gets a string from memory. */
 function getStringImpl(buffer, ptr) {
-  const len = new Uint32Array(buffer)[ptr + SIZE_OFFSET >>> 2] >>> 1;
-  const arr = new Uint16Array(buffer, ptr, len);
-  if (len <= STRING_DECODE_THRESHOLD) {
-    return String.fromCharCode.apply(String, arr);
+  let len = new Uint32Array(buffer)[ptr + SIZE_OFFSET >>> 2] >>> 1;
+  const wtf16 = new Uint16Array(buffer, ptr, len);
+  if (len <= STRING_SMALLSIZE) return String.fromCharCode.apply(String, wtf16);
```
Comment (MaxGraey) [inline lib/loader/index.js:48]: It looks like make sense cache `String.fromCharCode.apply` to constant of function. like:
```ts
const fromCharCodes = String.fromCharCode.apply;
```

Comment (dcodeIO) [reply lib/loader/index.js:48]: First will give you "apply is called on undefined". Second may or may not make a difference after gzip, hm.

Comment (MaxGraey) [reply lib/loader/index.js:48]: > Second may or may not make a difference after gzip, hm.

Right, ok

**Pull request merged**